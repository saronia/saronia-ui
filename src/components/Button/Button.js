/* @flow */
import R from 'ramda';
import React from 'react';
import styled from 'styled-components';
import type { Breed as BreedType, Magnitude as MagnitudeType } from '../../types';
import { Animation, Color, Font, Border, Breed, Magnitude } from '../../assets/constants';

type Props =
  | {|
      accent?: boolean,
      inverted?: boolean,
      onClick?: Function,
      pop?: 'active' | 'focus' | 'hover',
      breed?: BreedType,
      size?: MagnitudeType,
    |}
  | {|
      circular: true,
      radius: MagnitudeType,
      accent?: boolean,
      inverted?: boolean,
      onClick?: Function,
      pop?: 'active' | 'focus' | 'hover',
      pulse?: boolean,
      breed?: BreedType,
      size?: MagnitudeType,
    |};

const getBackgroundColor = R.cond([
  [R.propEq('inverted', true), R.always('transparent')],
  [R.propEq('breed', Breed.SECONDARY), R.always(Color.SECONDARY)],
  [R.T, R.always(colors.primary)],
]);

const getBackgroundHoverColor = R.cond([
  [R.propEq('inverted', true), R.always(colors.whiteDark)],
  [R.propEq('secondary', true), R.always(colors.secondaryDark)],
  [R.T, R.always(colors.primaryDark)],
]);

const getBorder = R.ifElse(
  props => R.isNil(getBorderColor(props)),
  R.always('none'),
  props => R.always(`${getBorderWidth(props)} solid ${getBorderColor(props)}`),
);

const getBorderColor = R.cond([
  [
    R.both(R.propEq('inverted', true), R.propEq('secondary', true)),
    R.always(colors.secondary),
  ],
  [R.propEq('inverted', true), R.always(colors.primary)],
  [R.T, R.always(undefined)],
]);

const getBorderWidth = R.cond([
  [R.propEq('inverted', 'thin'), R.always(borders.width.thin)],
  [R.propEq('inverted', 'medium'), R.always(borders.width.medium)],
  [R.propEq('inverted', 'thick'), R.always(borders.width.thick)],
  [R.T, R.always(borders.width.thin)],
]);

const getColor = R.cond([
  [
    R.both(R.propEq('inverted', true), R.propEq('secondary', true)),
    R.always(colors.secondary),
  ],
  [R.propEq('inverted', true), R.always(colors.primaryDark)],
  [R.propEq('secondary', true), R.always(colors.ivoryDark)],
  [R.T, R.always(colors.black)],
]);

const getFontSize = props => fonts.sizes[R.prop('size', props)];

const getPadding = R.cond([
  [R.propEq('radius', 'mini'), R.always(0)],
  [R.T, R.always('0.2rem 0.5rem')],
]);

const getTransform = R.curry((pseudoState, props) =>
  R.when(R.propEq('pop', pseudoState), R.always('scale(1.13)')),
);

const getWidth = R.cond([
  [R.propEq('radius', 'mini'), R.always('1.5rem')],
  [R.propEq('radius', 'tiny'), R.always('2rem')],
  [R.propEq('radius', 'small'), R.always('2.5rem')],
  [R.propEq('radius', 'medium'), R.always('3rem')],
  [R.propEq('radius', 'large'), R.always('4.5rem')],
  [R.propEq('radius', 'big'), R.always('6rem')],
  [R.propEq('radius', 'huge'), R.always('9rem')],
  [R.propEq('radius', 'massive'), R.always('12rem')],
  [R.T, R.always('inherit')],
]);

const Button = styled.button`
  background-color: ${getBackgroundColor};
  width: ${getWidth};
  height: ${getWidth};
  padding: ${getPadding};
  overflow: hidden;

  ${/* flexbox */ ''}
  display: flex;
  justify-content: ${props => (props.circular ? 'center' : 'flex-start')};

  ${/* font */ ''}
  color: ${getColor};
  font-size: ${getFontSize};
  font-family: ${props => (props.accent ? fonts.accent : fonts.system)};

  ${/* border */ ''}
  border: ${getBorder};
  border-radius: ${props => (props.circular ? '50%' : borders.radius)};

  ${/* transition */ ''}
  transition: all .3s, transform .4s ease-out;

  &:hover {
    cursor: pointer;
    background-color: ${getBackgroundHoverColor};
    transform: ${getTransform('hover')};
  }

  &:active {
    transform: ${getTransform('active')};
  }

  &:focus {
    transform: ${getTransform('focus')};
    outline: none;
  }
`;

const Pulse = styled.div`
  position: relative;
`;

const Overlay = styled(Button)`
  position: absolute;
  top: 0;
  left: 0;
  color: transparent;
  z-index: -1;
  animation-name: ${animations.scaleUpAndFadeOut};
  animation-duration: 1.6s;
  animation-iteration-count: infinite;
  animation-delay: .2s;
`;

export default (props: Props) => {
  return R.cond([
    [
      R.propEq('pulse', true),
      R.always(
        <Pulse>
          <Button {...props} />
          <Overlay {...R.merge(props, { inverted: true })} />
        </Pulse>,
      ),
    ],
    [R.T, R.always(<Button {...props} />)],
  ])(props);
};
