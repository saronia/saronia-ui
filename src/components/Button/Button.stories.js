/* @flow */
import R from 'ramda';
import React from 'react';
import { storiesOf, action, linkTo } from '@kadira/storybook';
import {
  withKnobs,
  text,
  boolean,
  number,
  select,
} from '@kadira/storybook-addon-knobs';

import { Button } from '.';
import { SizeEnum } from '../../assets/constants';

const stories = storiesOf('Button', module);

stories.addDecorator(withKnobs);
stories.addDecorator(story => (
  <div style={{ margin: '50px' }}>
    {story()}
  </div>
));

const popOptions = R.invertObj({
  ACTIVE: 'active',
  FOCUS: 'focus',
  HOVER: 'hover',
  NONE: undefined,
});

stories.add('default', () => {
  const circular = boolean('circular', false);
  const props = {
    secondary: boolean('secondary', false),
    accent: boolean('accent', false),
    inverted: boolean('inverted', false),
    size: select('size', R.invertObj(SizeEnum), 'medium'),
    children: text('children', 'Register'),
    circular,
    radius: circular
      ? select('radius', R.invertObj(SizeEnum), 'large')
      : undefined,
    pop: select('pop', popOptions, undefined),
    pulse: circular ? boolean('pulse', false) : undefined,
  };

  return R.cond([
    [
      R.propEq('circular', true),
      R.always(
        <Button
          accent={props.accent}
          circular
          inverted={props.inverted}
          onClick={action('clicked')}
          pop={props.pop}
          pulse={props.pulse}
          radius={props.radius || 'mini'}
          secondary={props.secondary}
          size={props.size}
        >
          {props.children}
        </Button>,
      ),
    ],
    [
      R.T,
      R.always(
        <Button
          accent={props.accent}
          inverted={props.inverted}
          onClick={action('clicked')}
          pop={props.pop}
          secondary={props.secondary}
          size={props.size}
        >
          {props.children}
        </Button>,
      ),
    ],
  ])(props);
});
