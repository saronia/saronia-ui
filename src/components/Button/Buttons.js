/* @flow */
import React, { Children, Component } from 'react';
import styled from 'styled-components';
import { BORDER_RADIUS } from '../../lib/constants';
import type { Breed, Magnitude } from '../../lib/types';

type Props = {
  accent?: boolean,
  breed?: Breed,
  children: Children,
  inverted?: boolean,
  size?: Magnitude,
  vertical?: boolean,
};

const Wrapper = styled.div`
  display: inline-flex;
  flex-direction: ${props => (props.vertical && 'column') || 'row'};

  button {
    border-radius: 0;

    &:not(:first-child):not(:last-child) {
      border-left: none;
      border-right: none;
    }

    &:first-child {
      border-top-left-radius: ${BORDER_RADIUS};
      border-top-right-radius: ${p => (p.vertical ? BORDER_RADIUS : 0)};
      border-bottom-left-radius: ${p => (p.vertical ? 0 : BORDER_RADIUS)};
    }

    &:last-child {
      border-top-right-radius: ${p => (p.vertical ? 0 : BORDER_RADIUS)};
      border-bottom-right-radius: ${BORDER_RADIUS};
      border-bottom-left-radius: ${p => (p.vertical ? BORDER_RADIUS : 0)};
    }
  }
`;

class Buttons extends Component<void, Props, void> {
  render() {
    return (
      <Wrapper {...this.props}>
        {React.Children.map(this.props.children, child =>
          React.cloneElement(child, {
            accent: this.props.accent,
            breed: this.props.breed,
            inverted: this.props.inverted,
            size: this.props.size,
          }),
        )}
      </Wrapper>
    );
  }
}

export default Buttons;
