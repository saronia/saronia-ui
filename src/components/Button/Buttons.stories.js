/* @flow */
import React from 'react';
import R from 'ramda';
import styled from 'styled-components';

import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import {
  withKnobs,
  text,
  boolean,
  number,
  select,
} from '@storybook/addon-knobs';
import centered from '@storybook/addon-centered';

import { Buttons, Button } from '.';
import { BreedEnum } from '../../lib/constants';

const breedOptions = R.invertObj(R.merge(BreedEnum, { DEFAULT: undefined }));
const isNotNil = R.both(
  R.complement(R.isNil),
  R.complement(R.equals('undefined')),
);

const stories = storiesOf('Buttons', module);
stories.addDecorator(withKnobs).addDecorator(centered);

stories.add('default', () => {
  const props = R.pickBy(isNotNil, {
    accent: boolean('accent', false),
    inverted: boolean('inverted', true),
    breed: select('breed', breedOptions, 'primary'),
  });

  return (
    <Buttons {...props} onClick={action('clicked')}>
      <Button>Login</Button>
      <Button>Do nothing</Button>
      <Button>Sign up</Button>
    </Buttons>
  );
});
