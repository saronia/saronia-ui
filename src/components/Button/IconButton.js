/* @flow */
import React from 'react';
import R from 'ramda';
import styled from 'styled-components';

import RawButton, {
  getBackgroundColorActive,
  getBackgroundColorHover,
} from './RawButton';
import { Icon } from '../Icon';
import { BreedEnum, MagnitudeEnum, Color } from '../../lib/constants';
import type { Props } from './Button';

const getPadding = R.cond([
  [R.propEq('radius', MagnitudeEnum.MINI), R.always(0)],
  [R.propEq('size', MagnitudeEnum.MINI), R.always('.2rem .3rem')],
  [R.propEq('size', MagnitudeEnum.TINY), R.always('.2rem .3rem')],
  [R.propEq('size', MagnitudeEnum.SMALL), R.always('.2rem .4rem')],
  [R.propEq('size', MagnitudeEnum.MEDIUM), R.always('.25rem .5rem')],
  [R.propEq('size', MagnitudeEnum.LARGE), R.always('.3rem .7rem')],
  [R.propEq('size', MagnitudeEnum.BIG), R.always('.35rem .75rem')],
  [R.propEq('size', MagnitudeEnum.HUGE), R.always('.4rem .9rem')],
  [R.propEq('size', MagnitudeEnum.MASSIVE), R.always('.5rem 1.2rem')],
  [R.T, R.always(R.always('.3rem .7rem'))],
]);

const getIconBackgroundColor = R.cond([
  [
    R.both(R.propEq('inverted', true), R.propEq('breed', BreedEnum.PRIMARY)),
    R.always(Color.PRIMARY),
  ],
  [
    R.both(R.propEq('inverted', true), R.propEq('breed', BreedEnum.SECONDARY)),
    R.always(Color.SECONDARY),
  ],
  [R.propEq('breed', BreedEnum.PRIMARY), R.always(Color.Primary.DARK)],
  [R.propEq('breed', BreedEnum.SECONDARY), R.always(Color.Secondary.DARK)],
  [R.T, R.always(Color.Gray.MODERATE)],
]);

const getIconFill = R.cond([
  [
    R.both(R.propEq('inverted', true), R.propEq('breed', BreedEnum.PRIMARY)),
    R.always(Color.WHITE),
  ],
  [
    R.both(R.propEq('inverted', true), R.propEq('breed', BreedEnum.SECONDARY)),
    R.always(Color.WHITE),
  ],
]);

const getIconFillHover = R.cond([
  [
    R.both(R.propEq('inverted', true), R.propEq('breed', BreedEnum.PRIMARY)),
    R.always(Color.WHITE_DARK),
  ],
  [
    R.both(R.propEq('inverted', true), R.propEq('breed', BreedEnum.SECONDARY)),
    R.always(Color.WHITE_DARK),
  ],
]);

const getIconFillActive = R.cond([
  [
    R.both(R.propEq('inverted', true), R.propEq('breed', BreedEnum.PRIMARY)),
    R.always(Color.IVORY_DARK),
  ],
  [
    R.both(R.propEq('inverted', true), R.propEq('breed', BreedEnum.SECONDARY)),
    R.always(Color.IVORY_DARK),
  ],
]);

const Wrapper = RawButton.withComponent('div').extend`
  padding: 0;
  align-items: stretch;

  svg {
    fill: ${getIconFill};
  }

  &:hover {
    svg {
      fill: ${getIconFillHover};
    }

    span {
      background-color: ${getBackgroundColorHover};
    }
  }

  &:active {
    span {
      background-color: ${getBackgroundColorActive};
    }

    svg {
      fill: ${getIconFillActive};
    }
  }
`;

const IconWrapper = styled.span`
  background-color: ${getIconBackgroundColor};
  padding: ${getPadding};
  display: flex;
  justify-content: center;
  align-items: center;
`;

const TextWrapper = styled.span`
  padding: ${getPadding};
  display: flex;
  justify-content: center;
  align-items: center;
`;

const IconButton = (props: Props) => {
  const passthroughProps = R.omit(['onClick', 'iconLeft', 'iconRight'], props);

  return (
    <Wrapper {...props}>
      {props.iconLeft &&
        <IconWrapper {...passthroughProps}>
          <Icon svgPath={props.iconLeft} inverted={!props.inverted} />
        </IconWrapper>}
      <TextWrapper {...passthroughProps}>
        {R.prop('children', props)}
      </TextWrapper>
      {props.iconRight &&
        <IconWrapper {...passthroughProps}>
          <Icon svgPath={props.iconRight} inverted={!props.inverted} />
        </IconWrapper>}
    </Wrapper>
  );
};

export default IconButton;
