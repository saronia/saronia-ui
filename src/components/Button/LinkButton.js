/* @flow */
import React from 'react';

import RawButton from './RawButton';
import type { Props } from './Button';

const LinkButton = RawButton.withComponent('a');

export default (props: Props) => <LinkButton {...props} />;
