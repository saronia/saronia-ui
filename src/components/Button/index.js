/* @flow */
import R from 'ramda';
import React from 'react';
import styled from 'styled-components';

import { colors, fonts, borders } from '../../assets/styles';

type Props = {
  inverted?: boolean | 'thin' | 'medium' | 'thick',
  accent?: boolean,
  secondary?: boolean,
};

const getBackgroundColor = R.cond([
  [R.prop('inverted'), R.always(colors.transparent)],
  [R.prop('secondary'), R.always(colors.secondary)],
  [R.T, R.always(colors.primary)],
]);

const getBackgroundHoverColor = R.cond([
  [R.prop('inverted'), R.always(colors.whiteDark)],
  [R.prop('secondary'), R.always(colors.secondaryDark)],
  [R.T, R.always(colors.primaryDark)],
]);

const getBorderWidth = R.cond([
  [R.propEq('inverted', 'thin'), R.always(borders.width.thin)],
  [R.propEq('inverted', 'medium'), R.always(borders.width.medium)],
  [R.propEq('inverted', 'thick'), R.always(borders.width.thick)],
  [R.T, R.always(borders.width.thin)],
]);

const getBorderColor = R.cond([
  [
    R.both(R.prop('inverted'), R.prop('secondary')),
    R.always(colors.secondary),
  ],
  [R.prop('inverted'), R.always(colors.primary)],
  [R.T, R.always(undefined)],
]);

const getBorder = R.ifElse(
  props => R.isNil(getBorderColor(props)),
  R.always('none'),
  props => R.always(`${getBorderWidth(props)} solid ${getBorderColor(props)}`),
);

const getColor = R.cond([
  [R.both(R.prop('inverted'), R.prop('secondary')), R.always(colors.secondary)],
  [R.prop('secondary'), R.always(colors.ivoryDark)],
  [R.T, R.always(colors.black)],
]);

const Button = styled.button`
  padding: ${props => props.accent ? '0.65rem 1rem 0.35rem 1rem' : '0.5rem 1rem'};
  background-color: ${getBackgroundColor};
  font-size: 1.2rem;
  font-family: ${props => props.accent ? fonts.accent : fonts.system};
  color: ${getColor};
  transition-duration: 0.25s;

  border: ${getBorder};
  border-radius: ${borders.radius};

  &:hover {
    cursor: pointer;
    background-color: ${getBackgroundHoverColor};
  }

  &:focus {
    outline: none;
  }
`;

export default (props: Props) => <Button {...props} />;
