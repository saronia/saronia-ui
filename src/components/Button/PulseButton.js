/* @flow */
import React from 'react';
import R from 'ramda';
import styled from 'styled-components';

import { Animation } from '../../lib/constants';
import type { Breed, Magnitude } from '../../lib/types';
import Button from './Button';

type Props = {
  accent?: boolean,
  circular: true,
  compact?: false,
  icon?: false,
  inverted?: boolean,
  onClick?: Function,
  pop?: 'active' | 'focus' | 'hover',
  radius: Magnitude,
  size?: Magnitude,
  breed?: Breed,
};

const Pulse = styled.div`
  position: relative;
`;

const Overlay = styled(Button)`
  position: absolute;
  top: 0;
  left: 0;
  color: transparent;
  z-index: -1;
  animation-name: ${Animation.SCALE_UP_AND_FADE_OUT};
  animation-duration: 1.6s;
  animation-iteration-count: infinite;
  animation-delay: .2s;
`;

const PulseButton = (props: Props) => (
  <Pulse>
    <Button {...props} />
    <Overlay {...R.merge(props, { inverted: true })} />
  </Pulse>
);

export default PulseButton;
