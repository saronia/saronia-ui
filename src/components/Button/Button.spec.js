/* @flow */
import React from 'react';
import renderer from 'react-test-renderer';
import { matcher, serializer } from 'jest-styled-components';

import Button from '.';

expect.addSnapshotSerializer(serializer);
expect.extend(matcher);

describe('<Button />', () => {
  const tree = renderer.create(<Button>Hello</Button>).toJSON();

  it('renders a button w/o crashing', () => {
    expect(tree.type).toEqual('button');
  });

  it('can render in an inverted style', () => {
    expect(tree).toMatchStyledComponentsSnapshot();
  });
});
