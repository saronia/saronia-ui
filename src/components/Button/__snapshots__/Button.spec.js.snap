// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> matches styled component snapshot 1`] = `
.keCvcC {
  background-color: #FFB00C;
  width: inherit;
  height: inherit;
  padding: 0.2rem 0.5rem;
  overflow: hidden;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  justify-content: flex-start;
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: flex-start;
  color: #161616;
  font-size: ;
  font-family: 'Quicksand', sans-serif;
  border: none;
  border-radius: 3px;
  transition: all .3s, -ms-transform .4s ease-out, -webkit-transform .4s ease-out, transform .4s ease-out;
  -webkit-transition: all .3s, -webkit-transform .4s ease-out, transform .4s ease-out;
}

.keCvcC:hover {
  cursor: pointer;
  background-color: #FFA104;
  transform: children:;
  -ms-transform: children:;
  -webkit-transform: children:;
  -hello: ;
}



.keCvcC:active {
  transform: children:;
  -ms-transform: children:;
  -webkit-transform: children:;
  -hello: ;
}



.keCvcC:focus {
  transform: children:;
  -ms-transform: children:;
  -webkit-transform: children:;
  -hello: ;
  outline: none;
}



<button
  className="keCvcC"
>
  Hello
</button>
`;
